### PR TITLE
feat: [#577] [#1170] Implement new Clock API

### PR DESCRIPTION
**Pull request by eonarheim**
===:clipboard: PR Checklist :clipboard:===

- [x] :pushpin: issue exists in github for these changes
- [x] :microscope: existing tests still pass
- [x] :see_no_evil: code conforms to the [style guide](https://github.com/excaliburjs/Excalibur/blob/main/STYLEGUIDE.md)
- [x] :triangular_ruler: new tests written and passing / old tests updated with new scenario(s)
- [x] :page_facing_up: changelog entry added (or not needed)

==================

Closes: #1170 
Closes: #577 

This PR implements a new Clock api to manage the core main loop. Clocks hide the implementation detail of how the mainloop runs, users just knows that it ticks somehow. Clocks additionally encapsulate any related browser timing, like `performance.now()`

1. `StandardClock` encapsulates the existing `requestAnimationFrame` api logic
2. `TestClock` allows a user to manually step the mainloop, this can be useful for frame by frame debugging #1170 
3. The base abstract clock implements the specifics of elapsed time 

* The new `maxFps` option is added to the engine constructor options. This is achieved by exiting the mainloop early in clock if the current elapsed time does not match the desired fps. This can be useful when you want to deliver a consistent experience across devices.

## Some fixed bugs that were illuminated as part of this
- Excalibur's initialize flow was flawed, this PR refactors the start() to mean "ready for game code to run", before it actually meant resources loaded and player clicked play
- Play button wasn't positioned until first update, now positioned on init
- Systems were not initialized until first update, now initialized with the scene

### After this PR (consolidate init before start resolves)

1. game.start(loader)
2. clock started
3. load resources
4. compute resolution 
5. run engine/scene/ecs system initialize
6. emit ready
7. start the game

### Before

1. game.start(loader);
2. mainloop raf started
3. load resources
4. run some initialize logic
5. compute resolution
4. start game
5. run initialize logic

## Changes:

- New Clock API for managing the mainloop
- Added `maxFps` engine option to constrain to a maximum fps 
- Updated fps sampling to be more useful
- Updates the tests to use the new TestClock (fun side-effect tests are ~20 seconds faster in CI!)
- Some small bug fixes that are related to clock timings

### CODE REVIEW

**Comment by eonarheim** [comment]
I want to have the tests moved over to the TestClock before merging, should speed up the test runs

File: src/engine/Loader.ts [hunk 1]
```diff
@@ -240,10 +240,11 @@ export class Loader extends Class implements Loadable<Loadable<any>[]> {
   /**
    * Shows the play button and returns a promise that resolves when clicked
    */
-  public showPlayButton(): Promise<void> {
+  public async showPlayButton(): Promise<void> {
     if (this.suppressPlayButton) {
       this.hidePlayButton();
-      return Promise.resolve();
+      // Delay is to give the logo a chance to show, otherwise don't delay
+      await delay(500, this._engine?.clock);
```
Comment (eonarheim) [inline src/engine/Loader.ts:247]: This is a refactoring moving delay from engine to the loader, it doesn't change the behavior other than using the clock

File: src/engine/Loader.ts [hunk 2]
```diff
@@ -258,7 +259,8 @@ export class Loader extends Class implements Loadable<Loadable<any>[]> {
           this._playButton.click();
         }
       });
-      const promise = new Promise<void>((resolve) => {
+      this._positionPlayButton();
```
Comment (eonarheim) [inline src/engine/Loader.ts:262]: Small bug discovered, the button isn't positioned until the first update, this makes sure it's in the right position immediately

File: src/engine/tsconfig.json [hunk 3]
```diff
@@ -4,7 +4,7 @@
     "moduleResolution": "node",
     "sourceMap": true,
     "experimentalDecorators": true,
-    "target": "es2015",
+    "target": "es2018",
```
Comment (eonarheim) [inline src/engine/tsconfig.json:7]: Might move this to another pr, but we can safely target 2018, maybe higher?

File: src/engine/Util/Clock.ts [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+  public toStandardClock() {
+    const clock = new StandardClock({
+      ...this._options
+    });
+    return clock;
+  }
+
+  /**
+   * Schedule a callback to fire given a a timeout in milliseconds using the excalibur [[Clock]]
```
Comment (jedeen) [inline src/engine/Util/Clock.ts:82]: "given a timeout"

File: src/engine/Util/Clock.ts [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+   *
+   * This is useful to use over the built in browser `setTimeout` because callbacks will be tied to the
+   * excalibur update clock, instead of browser time, this means that callbacks wont fire if the game is
+   * stopped or paused.
+   *
+   * @param cb callback to fire
+   * @param timeoutMs Optionally sepcify a timeout in milliseconds from now, default is 0ms which means the next possible tick
```
Comment (jedeen) [inline src/engine/Util/Clock.ts:89]: "specify"

**Review by jedeen — APPROVED** [review]
